skopt/trees.py

```diff
@@ -0,0 +1,139 @@
+"""Gradient boosted trees based minimization algorithms."""
+
+import numpy as np
+
+from scipy import stats
+from scipy.optimize import fmin_l_bfgs_b
+from scipy.optimize import OptimizeResult
+
+from sklearn.base import clone
+from sklearn.utils import check_random_state
+
+from .gbt import GradientBoostingQuantileRegressor
+from .utils import extract_bounds
+
+
+def _expected_improvement(X, surrogate, best_y):
+    """Evaluate expected improvement for `surrogate` model at `x`"""
+    X = np.asarray(X)
+    if X.ndim == 1:
+        X = np.expand_dims(X, axis=0)
+
+    # low and high are assumed to be the 16% and 84% quantiles
+    low, mu, high = surrogate.predict(X)
+    # approximate the std dev, if the pdf is gaussian this is exact
+    std = (high - low) / 2.
+
+    ei = np.zeros(len(mu))
+
+    mask = std > 0
+    improvement = best_y - mu[mask]
+    exploit = improvement * stats.norm.cdf(improvement / std[mask])
+    explore = std[mask] * stats.norm.pdf(improvement / std[mask])
+    ei[mask] = exploit + explore
+
+    # do not want EI below zero, happens for points that are (much)
+    # worse than ``best_y``
+    ei = np.clip(ei, 0., np.max(ei))
+
+    ei = -ei # we are being used in a minimizer
+    return ei
+
+
+def _random_point(lower, upper, random_state=None):
+    """Sample a random point"""
+    rng = check_random_state(random_state)
+
+    num_params = len(lower)
+    delta = upper - lower
+    return lower + rng.rand(num_params) * delta
+
+
+def gbt_minimize(func, bounds, base_estimator=None, maxiter=100,
```

Comment:
also, as part of #38, I added an `n_start` argument. Surely, this should be helpful here as well.